cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain/Properties.hs

```diff
@@ -0,0 +1,158 @@
+import Data.Typeable
+import Shelley.Spec.Ledger.API.Genesis
+import Test.Cardano.Ledger.ModelChain
+import Test.Cardano.Ledger.ModelChain.Utils
+import Test.Tasty
+import Test.Tasty.QuickCheck
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Val as Val
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import Data.Ratio ((%))
+import qualified Cardano.Ledger.Era
+
+import Cardano.Ledger.Shelley (ShelleyEra)
+import Cardano.Ledger.Alonzo (AlonzoEra)
+import Test.Cardano.Ledger.Elaborators.Shelley ()
+import Test.Cardano.Ledger.Elaborators.Alonzo ()
+import Test.Shelley.Spec.Ledger.ConcreteCryptoTypes (C_Crypto)
+
+newTestFw ::
+  forall era proxy.
+  ( ElaborateEraModel era
+  , HasEraElaboratorState (ElaborateEraModelState era) era
+  , Default (AdditionalGenesisConfig era)
+  , Default (ElaborateEraModelState era)
+  , Eq (PredicateFailure (Core.EraRule "LEDGER" era))
+  , Show (PredicateFailure (Core.EraRule "LEDGER" era))
+  , Cardano.Ledger.Era.Era era
+  )
+  => proxy era -> TestTree
+newTestFw proxy = testGroup (show $ typeRep proxy)
+  [ testProperty "noop" $ testChainModelInteraction proxy Map.empty []
+
+  , testProperty "noop-2" $ testChainModelInteraction proxy
+    (Map.fromList
+      [ ("alice", Coin 1_000_000)
+      , ("bob", Coin 1_000_000)
+      ])
+    [ModelEpoch [] mempty]
+  , let
+      genAct = Map.fromList
+        [ ("alice", Coin 1_000_000_000_000)
+        ]
+      checks nes ems =
+        let rewards = observeRewards (nes, ems)
+        in counterexample (show rewards) $ Coin 0 === fold rewards
+
+    in testProperty "deleg" $ testChainModelInteractionWith proxy checks genAct
+      [ ModelEpoch
+        [ ModelBlock 0
+          [ (modelTx 1)
+            { _mtxInputs = Set.fromList [ModelGensisIn "alice"]
+            , _mtxOutputs =
+              [ ModelTxOut "alice" ( 1_000_000_000_000 - ( 100_000_000_000))
+              ]
+            , _mtxFee = 100_000_000_000
+            , _mtxDCert =
+              [ ModelRegisterStake "alice"
+              , ModelRegisterPool (ModelPoolParams "pool1" (Coin 0) (Coin 0) (unitIntervalFromRational (0 % 1) ) "alice" ["alice"])
+              , ModelDelegate (ModelDelegation "alice" "pool1")
+              ]
+            }
+          ]
+        ] (ModelBlocksMade $ Map.fromList [])
+      , ModelEpoch [] (ModelBlocksMade $ Map.fromList [])
+      , ModelEpoch [] (ModelBlocksMade $ Map.fromList [("pool1", 100)])
+      , ModelEpoch [] (ModelBlocksMade $ Map.fromList [])
+      , ModelEpoch
+        [ ModelBlock 0
+          [ (modelTx 1)
+            { _mtxInputs = Set.fromList [ModelTxIn 1 0]
+            , _mtxOutputs =
+              [ ModelTxOut "alice" ( 1_000_000_000_000 - ( 2 * 100_000_000_000))
+              , ModelTxOut "bob" 8999820000000
+              ]
+            , _mtxFee = 100_000_000_000
+            , _mtxWdrl = Map.singleton "alice" (Coin 8999820000000)
```

Comment:
I'm confused what's going on here. I see Bob getting `8999820000000` in an output. And that same number used for what look like a reward withdrawal for Alice. Are these related, and how do you know how much Alice can withdraw?

Reply:
i guess the way to interpret this is that "alice" is gifting "bob" with the rewards, the addresses used aren't meaningfully distinct entities, only distinct keys.  There's no significance to the addresses being different.

The value is what happens to be rewarded in this exact sequence of inputs.  The hardcoded value goes away in another branch ( https://github.com/obsidiansystems/cardano-ledger-specs/compare/model-based-testing...obsidiansystems:model-based-testing-ma?expand=1 )

Reply:
ah!